<a href="https://colab.research.google.com/github/y-fares/PSTB-YCDynamics-Hackathon-1/blob/DEV_YACINE/yc_dynamics_hackathon1_subj3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

---
---

**PSTB Bootcamp - Hackathon 1 - Subject 3**: *Exploring Employee Attrition and Performance in a Corporate Environment (IBM HR Analytics)*


---


**Date** : 23/10/2025


---


**Team**: YC Dynamics (Yacine Fares, Clément Philbert)

---
---

# 1 ) Data Loading and Preprocessing

## a. Downloading the IBM HR Analytics dataset and extracting the CSV

In [60]:
## Downloading the IBM HR Analytics dataset using the requests module (version 2.32.5)

# Importing the requests library
import requests
url = 'https://github.com/devtlv/Datasets-DA-Bootcamp-2-/raw/refs/heads/main/Week%208%20-%20Hackathon%20&%20AB%20Testing/W8D1%20-%20Hackathon/IBM%20HR%20Analytics%20Employee%20Attrition%20&%20Performance.zip'

response = requests.get(url)
file_Path = 'IBM_HR_Analytics_Employee_Attrition_Performance.zip'

if response.status_code == 200:
    with open(file_Path, 'wb') as file:
        file.write(response.content)
    print('File downloaded successfully')
else:
    print('Failed to download file')

File downloaded successfully


In [61]:
## Extracting the IBM HR Analytics csv file from the archive and moving it to a "dataset" folder using
## the zipfile and os modules (dropping unnecessary folder cascading from the zip).

# Importing libraries
import zipfile
import os

zip_path = "IBM_HR_Analytics_Employee_Attrition_Performance.zip"
file_to_extract = "IBM HR Analytics Employee Attrition & Performance/IBM HR Analytics Employee Attrition & Performance/WA_Fn-UseC_-HR-Employee-Attrition.csv"   # the exact internal path
output_dir = "dataset"

# Making sure the output folder exists
os.makedirs(output_dir, exist_ok=True)

with zipfile.ZipFile(zip_path, 'r') as zip_ref:
    with zip_ref.open(file_to_extract) as source_file:
        # Get only the filename (drop folders)
        filename = os.path.basename(file_to_extract)
        output_path = os.path.join(output_dir, filename)

        # Write file contents
        with open(output_path, 'wb') as target_file:
            target_file.write(source_file.read())

print(f"Extracted {file_to_extract} → {output_path}")

Extracted IBM HR Analytics Employee Attrition & Performance/IBM HR Analytics Employee Attrition & Performance/WA_Fn-UseC_-HR-Employee-Attrition.csv → dataset/WA_Fn-UseC_-HR-Employee-Attrition.csv


## b. Loading the csv dataset into a dataframe with pandas

In [62]:
## Loading the IBM HR Analytics dataset in a dataframe

# Importing the pandas library
import pandas as pd

# Creating the dataframe from the csv
df = pd.read_csv("dataset/WA_Fn-UseC_-HR-Employee-Attrition.csv")

## c. Initial high-level data exploration

In [63]:
# Checking first five entries of the dataset with .head()
df.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

In [64]:
# Checking structure and datatypes with .info()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

> 👉 The dataset has 1470 entries and 35 columns

> 👉 Conclusions and recommendations regarding data transformation to perform:

| Columns currently int64                                                                                                                                                                                                | Current dtype | Suggested dtype                                | Reason                                                                                                              |
| -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------- | ---------------------------------------------- | ------------------------------------------------------------------------------------------------------------------- |
| `Age`                                                                                                                                                                                                    | int64         | ✅ keep as int64                                | Continuous numeric variable.                                                                                        |
| `DailyRate`, `HourlyRate`, `MonthlyIncome`, `MonthlyRate`                                                                                                                                                | int64         | ✅ keep as int64                                | Continuous numeric values.                                                                                          |
| `DistanceFromHome`, `PercentSalaryHike`, `YearsAtCompany`, `YearsInCurrentRole`, etc.                                                                                                                    | int64         | ✅ keep as int64                                | Continuous or count-based — fine as numeric.                                                                        |
| `Education`, `EnvironmentSatisfaction`, `JobInvolvement`, `JobLevel`, `JobSatisfaction`, `PerformanceRating`, `RelationshipSatisfaction`, `StockOptionLevel`, `TrainingTimesLastYear`, `WorkLifeBalance` | int64         | ⚠️ **convert to categorical (pd.Categorical)** | These are *ordinal* variables — coded as 1–4 or 1–5. Converting to categorical helps with readability and analysis. |
| `EmployeeCount`, `StandardHours`                                                                                                                                                                         | int64         | ⚠️ **drop or convert to constant**             | Usually constant across all rows (e.g., 1 or 80). They add no value — verify and drop if true.                      |
| `EmployeeNumber`                                                                                                                                                                                         | int64         | ⚠️ **treat as object/string (ID)**             | It’s an identifier, not numeric data — shouldn’t be used in math.                                                   |


| Columns currently object                                                                                             | Current dtype | Suggested dtype                       | Reason                                                                          |
| -------------------------------------------------------------------------------------------------- | ------------- | ------------------------------------- | ------------------------------------------------------------------------------- |
| `Attrition`                                                                                        | object        | ⚠️ **convert to boolean or category** | Typically “Yes” / “No”. Convert with `map({'Yes': 1, 'No': 0})` or to category. |
| `BusinessTravel`, `Department`, `EducationField`, `Gender`, `JobRole`, `MaritalStatus`, `OverTime` | object        | ⚠️ **convert to categorical**         | Nominal categorical variables — better memory and efficiency.                   |
| `Over18`                                                                                           | object        | ⚠️ **drop or set as constant**        | Usually all "Y". If so, drop (no information value).                            |


## d. Looking for missing values or duplicates

In [65]:
# Looking for potential missing values in the dataframe (returning the sum of null values per column)
print(df.isnull().sum())
print(f"\n\nNumber of missing data in datateset: ", df.isnull().sum().sum())

Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSince

> 👉 There are no missing values in the dataset

In [66]:
# Looking for potential duplicates in the dataframe (returning the count as an integer)
df.duplicated().sum()

np.int64(0)

> 👉 There are no duplicates in the dataset (returned value is 0)

## e. Converting and cleaning the dataset before analysis

Dropping irrelevant columns

In [67]:
df.drop(columns=['EmployeeCount', 'EmployeeNumber', 'Over18', 'StandardHours'], inplace=True)

In [68]:
# Verifying the columns dropped
df.shape

(1470, 31)

In [69]:
# Encoding categorial columns to simplify comparisons
#df['MaritalStatus'] = df['MaritalStatus'].map({"Single" : 0, "Married" : 1, "Divorced" : 2})
#df['OverTime'] = df['OverTime'].map({"No" : 0, "Yes" : 1})
#df['Gender'] = df['Gender'].map({"Female" : 0, "Male" : 1})
#df['Attrition'] = df['Attrition'].map({"No" : 0, "Yes" : 1})

In [70]:
df.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EnvironmentSatisfaction  \
0                 1          2  Life Sciences                        2   
1                 8          1  Life Sciences                        3   
2                 2          2          Other                        4   
3                 3          4  Life Sciences                        4   
4                 2          1        Medical                        1   

   Gender  ...  PerformanceRating  RelationshipSatisfaction  StockOptionLevel  \
0  Female  ...                  3                         1                 0   
1    Male  ...                  4                         4                 1   
2    Male  ...                  3                         2                 0   
3  Female  ...                  3                         3                 0   
4    Male  ...                  3                         4                 1   

  TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                 8                      0               1               6   
1                10                      3               3              10   
2                 7                      3               3               0   
3                 8                      3               3               8   
4                 6                      3               3               2   

   YearsInCurrentRole  YearsSinceLastPromotion YearsWithCurrManager  
0                   4                        0                    5  
1                   7                        1                    7  
2                   0                        0                    0  
3                   7                        3                    0  
4                   2                        2                    2  

[5 rows x 31 columns]

# 2 ) Exploratory Data Analysis (EDA)

In [71]:
# Percentage of Attrition
df["Attrition"].value_counts(normalize=True)

Attrition
No     0.838776
Yes    0.161224
Name: proportion, dtype: float64

To date, 16 % of employees have left the company

In [80]:
df.describe()

Age    DailyRate  DistanceFromHome    Education  \
count  1470.000000  1470.000000       1470.000000  1470.000000   
mean     36.923810   802.485714          9.192517     2.912925   
std       9.135373   403.509100          8.106864     1.024165   
min      18.000000   102.000000          1.000000     1.000000   
25%      30.000000   465.000000          2.000000     2.000000   
50%      36.000000   802.000000          7.000000     3.000000   
75%      43.000000  1157.000000         14.000000     4.000000   
max      60.000000  1499.000000         29.000000     5.000000   

       EnvironmentSatisfaction   HourlyRate  JobInvolvement     JobLevel  \
count              1470.000000  1470.000000     1470.000000  1470.000000   
mean                  2.721769    65.891156        2.729932     2.063946   
std                   1.093082    20.329428        0.711561     1.106940   
min                   1.000000    30.000000        1.000000     1.000000   
25%                   2.000000    48.000000        2.000000     1.000000   
50%                   3.000000    66.000000        3.000000     2.000000   
75%                   4.000000    83.750000        3.000000     3.000000   
max                   4.000000   100.000000        4.000000     5.000000   

       JobSatisfaction  MonthlyIncome  ...  PerformanceRating  \
count      1470.000000    1470.000000  ...        1470.000000   
mean          2.728571    6502.931293  ...           3.153741   
std           1.102846    4707.956783  ...           0.360824   
min           1.000000    1009.000000  ...           3.000000   
25%           2.000000    2911.000000  ...           3.000000   
50%           3.000000    4919.000000  ...           3.000000   
75%           4.000000    8379.000000  ...           3.000000   
max           4.000000   19999.000000  ...           4.000000   

       RelationshipSatisfaction  StockOptionLevel  TotalWorkingYears  \
count               1470.000000       1470.000000        1470.000000   
mean                   2.712245          0.793878          11.279592   
std                    1.081209          0.852077           7.780782   
min                    1.000000          0.000000           0.000000   
25%                    2.000000          0.000000           6.000000   
50%                    3.000000          1.000000          10.000000   
75%                    4.000000          1.000000          15.000000   
max                    4.000000          3.000000          40.000000   

       TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
count            1470.000000      1470.000000     1470.000000   
mean                2.799320         2.761224        7.008163   
std                 1.289271         0.706476        6.126525   
min                 0.000000         1.000000        0.000000   
25%                 2.000000         2.000000        3.000000   
50%                 3.000000         3.000000        5.000000   
75%                 3.000000         3.000000        9.000000   
max                 6.000000         4.000000       40.000000   

       YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
count         1470.000000              1470.000000           1470.000000  
mean             4.229252                 2.187755              4.123129  
std              3.623137                 3.222430              3.568136  
min              0.000000                 0.000000              0.000000  
25%              2.000000                 0.000000              2.000000  
50%              3.000000                 1.000000              3.000000  
75%              7.000000                 3.000000              7.000000  
max             18.000000                15.000000             17.000000  

[8 rows x 23 columns]

| Infos        | Age                | Distance From Home | Years at company | Years In Current Role | Years Since Last Promotion | Years With Current Manager |
|--------------|--------------------|--------------------|------------------|-----------------------|----------------------------|----------------------------|
| Average      | 37 years old       | 9 km               | 7 years          | 4.2                   | 2.2                        | 4.1                        |
| Over 75 %    | above 43 years old | 14 km away         | 9 years          | 7                     | 3                          | 7                          |
| Oldest / Max | 60 years old       | 29 km              | 40 years         | 18                    | 15                         | 17                         |

In [81]:
num_cols = df.select_dtypes(include=['int64','float64']).columns.tolist()
cat_cols = df.select_dtypes(include='object').columns.tolist()

print("\nVariables numériques :", len(num_cols))
print(num_cols)

print("\nVariables catégorielles :", len(cat_cols))
print(cat_cols)


Variables numériques : 23
['Age', 'DailyRate', 'DistanceFromHome', 'Education', 'EnvironmentSatisfaction', 'HourlyRate', 'JobInvolvement', 'JobLevel', 'JobSatisfaction', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked', 'PercentSalaryHike', 'PerformanceRating', 'RelationshipSatisfaction', 'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance', 'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager']

Variables catégorielles : 8
['Attrition', 'BusinessTravel', 'Department', 'EducationField', 'Gender', 'JobRole', 'MaritalStatus', 'OverTime']


In [83]:
print("\nMoyennes par groupe (attrition = 0 vs 1) :")
print(df.groupby("Attrition")[["Age","MonthlyIncome","DistanceFromHome","YearsAtCompany"]].mean())



Moyennes par groupe (attrition = 0 vs 1) :
                 Age  MonthlyIncome  DistanceFromHome  YearsAtCompany
Attrition                                                            
No         37.561233    6832.739659          8.915653        7.369019
Yes        33.607595    4787.092827         10.632911        5.130802


### EDA Summary (without data visualization)

- The dataset contains **1,470 rows and 35 columns**.  
- The **attrition rate is 16%**, indicating an **imbalanced dataset**.  
- There are **no missing values** in the data.  
- **Main numerical variables:** Age, Monthly Income, Distance from Home, Years at Company.  
- **Main categorical variables:** Job Role, Department, Marital Status, Gender, OverTime.  
- **Observed trend:** Employees who work overtime, earn lower salaries, or live farther from work tend to have higher attrition rates.  
- **Strong negative correlations:** MonthlyIncome, Age, YearsAtCompany.  
- **Positive correlations:** DistanceFromHome, OverTime.  


# 3 ) Correlation and Trend Analysis

# 4 ) Visualization

# 5 ) Summary of findings

# 6 ) Retention Strategy Development